### PR TITLE
Refactor Matrix Implementation

### PR DESCRIPTION
**Pull request by EmmaMunley**
# Changes

In this change, we refactor code implementing Matrix.Include, including:
- Updating the struct backing the Matrix.Include type from []MatrixInclude to IncludeParamsList
   - IncludeParamsList that allows us create member functions of Matrix.Include
- Exporting functions that will be used in subsequent PRs that implement Matrix.Include
- Removing ResolvedPipelineTask.IsMatrixed() and reusing PipelineTask.ResolvedPipelineTask()

There are no functional changes in this commit.

/kind Cleanup

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by jerop** [comment]
`TestHermeticTaskRun` failed but is unrelated

/retest

**Comment by EmmaMunley** [comment]
> @EmmaMunley could you add a bit more detail to the commit message on the "why" of this change? I don't quite understand why this will make future changes easier

@lbernick Sure just updated the commit message. Most of the changes the tests to reflect the updated struct, which is used to create member functions. Also found that there were still old []Params instead of Params being used in testing.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 1]
```diff
@@ -106,7 +106,7 @@ func (t ResolvedPipelineTask) IsCustomTask() bool {
 
 // IsMatrixed return true if the PipelineTask has a Matrix.
 func (t ResolvedPipelineTask) IsMatrixed() bool {
-	return t.PipelineTask.Matrix != nil && len(t.PipelineTask.Matrix.Params) > 0
+	return t.PipelineTask.Matrix.HasParams() || t.PipelineTask.Matrix.HasInclude()
```
Comment (jerop) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:109]: this definition is different from pt.IsMatrixed in v1beta1/v1 -- also not sure that we need to redefine the function here

```suggestion
	return t.PipelineTask.IsMatrixed()
```

**Review by jerop — CHANGES_REQUESTED** [review]
Thanks for doing the refactor separately @EmmaMunley!

**Comment by EmmaMunley** [comment]
Note: This PR is blocking the subsequent PRs:
https://github.com/tektoncd/pipeline/pull/6418
#6341

File: pkg/apis/pipeline/v1/param_types_test.go [hunk 2]
```diff
@@ -453,3 +454,41 @@ func TestArrayReference(t *testing.T) {
...
+	}, {
+		name: "extract param names from ParamTypeString and ParamTypeArray",
+		params: v1.Params{{
+			Name: "GOARCH", Value: v1.ParamValue{Type: v1.ParamTypeArray, ArrayVal: []string{"linux/amd64", "linux/ppc64le", "linux/s390x"}},
+		}, {
+			Name: "IMAGE", Value: v1.ParamValue{Type: v1.ParamTypeString, StringVal: "image-1"},
+		}},
+		want: sets.NewString("GOARCH", "IMAGE"),
+	}}
+	for _, tt := range tests {
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/param_types_test.go:489]: can you add a test for duplicate param names?
(TBH I'm sort of confused why you'd like to dedup, I would imagine duplicate params would be an error)

Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/param_types_test.go:489]: @lbernick Added a test for duplicates. Checking for duplicate parameters within a Matrix is handled in a separate function `validateParams()`. This function `validateParameterInOneOfMatrixOrParams()` is checking that there aren't duplicates between PT Parameters and Matrix Parameters, so we can ignore duplicates in Matrix Parameters.

Comment (EmmaMunley) [reply pkg/apis/pipeline/v1/param_types_test.go:497]: Indeed I did 🤦‍♀️
